File: 2022/19.py
print(chr(27)+'[2j')
print('\033c')
f = open('19.test', 'r')
f = open('19.input', 'r')
lines = [x.strip() for x in f.readlines()]
print("Day 19")
import copy
from itertools import permutations
#from functools import lru_cache

blueprints = []
for l in lines:
    b = l.split(': ')[1]
    r = b.split('. ')
    blueprint = {} 
    for robot in r:
        t = robot.split(' ')[1] 
        costs = robot.split(' costs ')[1].split(' and ')
        costs = tuple([cost.replace('.', '') for cost in costs])
        costs = [c.split() for c in costs] 
        blueprint[t] = [(int(c[0]), c[1]) for c in costs]
    blueprints.append(blueprint)

def inc_resources(resources, robots):
    new_resources = list(resources)
    for r, _ in enumerate(resources):
        new_resources[r] = resources[r] + 1 * robots[r]
    return tuple(new_resources)

def inc_robot(robots, robot_type):
    new_robots = list(robots) 
    new_robots[I[robot_type]] += 1
    return tuple(new_robots)

itr = 0
I = { 'ore': 0, 'clay': 1, 'obsidian': 2, 'geode': 3 }

def game(t, resources, robots):
    global blueprint, itr
    blueprint = blueprint if blueprint is not None else {}
    itr += 1

    # Can purchase
    purchase = []
    for b in ['geode', 'obsidian', 'clay', 'ore']:
        can_purchase = True
        for (cost, cost_type) in blueprint[b]:
            if cost <= resources[I[cost_type]]:
                continue
            can_purchase = False
            break
        if can_purchase:
            purchase.append(b)

    actions = []
    orig_robots = copy.deepcopy(robots)
    new_resources = inc_resources(resources, orig_robots)
    if 'geode' not in purchase:
        actions.append((t+1, new_resources, orig_robots))

    #for i in range(len(purchase)):
    #    perm = permutations(purchase, 1)
    if len(purchase) >= 2 and 'ore' in purchase:
        purchase = [purchase[0], 'ore']
    for p in purchase[:2]:
        res2 = [r for r in new_resources] 
        new_robots = copy.deepcopy(robots)
        #p = [p]
        #for r in p:
        r = p
        for cost, cost_type in blueprint[r]:
            res2[I[cost_type]] -= cost
        new_robots = inc_robot(new_robots, r)
        
        actions.append((t+1, tuple(res2), new_robots))
            
    return actions

priority = ["geode", "obsidian", "clay", "ore"]
blueprint = None
results = []
for i, b in enumerate(blueprints):
    blueprint = b

    q = []
    max_geode = 0
    max_obsid = 0
    max_clay = 0
    max_ore = 0
    max_rob = (0,0,0,0) 
    robots =  (1,0,0,0) 
    resources =  (0,0,0,0) 
    MAX_T = 24
    reach = {}
    lvl = 0

    def get_max(res, rob):
        global max_ore, max_clay, max_obsid, max_geode, lvl, max_rob
        if(res[0] >= max_ore):
            max_ore = res[0]
            if lvl == 0:
                max_rob = rob
        if(res[1] >= max_clay):
            max_clay = res[1]
            if lvl < 1:
                max_rob = rob
                lvl +=1
            if lvl == 1:
                max_rob = rob
        if(res[2] >= max_obsid):
            max_obsid = res[2]
            if lvl < 2:
                lvl +=2
                max_rob = rob
            if lvl == 2:
                max_rob = rob
        if(res[3] >= max_geode):
            max_geode = res[3]
            if lvl < 3:
                lvl +=3
            if lvl == 3:
                max_rob = rob

    for t in range(MAX_T+1):
        reach[t] = set() 
    reach[0].add((0, resources, robots))

    print("-"*20)
    print('Blueprint', i+1)
    for t in range(0, MAX_T):

        count_obsid = 0
        prev_max = max_geode
        for (t1, res1, rob1) in reach[t]:
            new_actions = game(t1, res1, rob1)
            for (t2, res2, rob2) in new_actions:
                reach[t2].add((t2, res2, rob2))
                get_max(res2, rob2)
                if res2[2] == 0:
                    count_obsid += 1
        print("%d reaches %d. Maxes at geode:%d, obsidian:%d, clay:%d, ore:%d. Count obsidian:%d" 
              % (t, len(reach[t]), max_geode, max_obsid, max_clay, max_ore, count_obsid)
              + ". Robots:", max_rob 
        )
        #if prev_max == 0 and max_geode != 0:
        #    reach[t] = [x for x in reach[t] if x[1][3] != 0]
        

    geodes = [res[3] for (_, res, _) in reach[len(reach)-1]]
    geodes.sort()
    print("Finished with blueprint", i, ", score: ", geodes[-1])
    res = (i+1) * geodes[-1]
    results.append(res)


print('-'*80)
print('FINISHED')
for i,v in enumerate(results):
    print("Blueprint %d: %d" % (i+1, v))

print("Total result:", sum(results))

File: 2022/19.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"
)

const MAX_T = 32

func timer(name string) func() {
	start := time.Now()
	return func() {
		fmt.Printf("%s took %v\n", name, time.Since(start))
	}
}

func inc_robot(blueprint [7]int, state [9]int, r int) [9]int {
	if r == 0 { // ORE
		state[1] += 1
		state[5] -= blueprint[1]
	} else if r == 1 { // CLAY
		state[2] += 1
		state[5] -= blueprint[2]
	} else if r == 2 { // OBSIDIAN
		state[3] += 1
		state[5] -= blueprint[3]
		state[6] -= blueprint[4]
	} else if r == 3 { // GEODE
		state[4] += 1
		state[5] -= blueprint[5]
		state[7] -= blueprint[6]
	}
	return state
}

func inc_resources(state [9]int) [9]int {
	state[5] += state[1]
	state[6] += state[2]
	state[7] += state[3]
	state[8] += state[4]
	return state
}

func game(blueprint [7]int, state [9]int) [][9]int {
	// Increase time
	state[0] = state[0] + 1
	new_states := [][9]int{}

	// -- Can purchase robot
	can_purchase := [4]bool{true, true, true, true}
	can_purchase[0] = blueprint[1] <= state[5]
	can_purchase[1] = blueprint[2] <= state[5]
	can_purchase[2] = blueprint[3] <= state[5] && blueprint[4] <= state[6]
	can_purchase[3] = blueprint[5] <= state[5] && blueprint[6] <= state[7]

	// Increase resources
	new_state := inc_resources(state)

	// Get 2 first purchases.
	purchase := []int{}
	for i, v := range can_purchase {
		if v {
			purchase = append(purchase, i)
		}
	}

	// Only purchase ORE with another thing simultanously
	//if len(purchase) > 2 {
	//	new_purchase := []int{}
	//	new_purchase = append(new_purchase, purchase[0])
	//	new_purchase = append(new_purchase, purchase[len(purchase)-1])
	//	purchase = new_purchase
	//}

	for _, p := range purchase {
		new_state2 := inc_robot(blueprint, new_state, p)
		new_states = append(new_states, new_state2)
	}

	// Purchase nothing (if one cannot buy geode)
	if !can_purchase[3] {
		new_states = append(new_states, new_state)
	}
	return new_states
}

// State values:
// 0: TIME
// 1: ROBOT ORE
// 2: ROBOT CLAY
// 3: ROBOT OBSIDIAN
// 4: ROBOT GEODE
// 5: RESOURCE ORE
// 6: RESOURCE CLAY
// 7: RESOURCE OBSIDIAN
// 8: RESOURCE GEODE

func use_blueprint(blueprint [7]int) int {
	defer timer("use_blueprint")()
	init_state := [9]int{0, 1, 0, 0, 0, 0, 0, 0, 0}
	max_geodes := [MAX_T]int{}

	levels := [MAX_T][][9]int{}

	t := 0
	levels[t] = append(levels[t], init_state)
	for {
		max_geode := 0
		start := time.Now()
		reach := map[[9]int]bool{}
		for _, nxt := range levels[t] {
			if nxt[0] > MAX_T {
				continue
			}
			_, ok := reach[nxt]
			if ok {
				continue
			}
			reach[nxt] = true
			new_states := game(blueprint, nxt)
			for _, state := range new_states {
				if t+1 < MAX_T {
					if t > 10 {
						if state[8] >= max_geodes[t-2] {
							levels[t+1] = append(levels[t+1], state)
						}
					} else {
						levels[t+1] = append(levels[t+1], state)
					}
				}
				if state[8] > max_geode {
					max_geode = state[8]
				}
			}
			max_geodes[t] = max_geode
		}
		fmt.Printf("T=%d took %v, with max: %d \n", t+1, time.Since(start), max_geodes[t])
		t = t + 1
		if t >= MAX_T {
			break
		}
	}

	fmt.Println(max_geodes)
	return max_geodes[MAX_T-1]
}

func main() {
	fmt.Println("Day 19")
	//content, err := os.ReadFile("19.test")
	content, err := os.ReadFile("19.input")
	if err != nil {
		panic(err)
	}
	re, err := regexp.Compile(`\d+`)
	if err != nil {
		panic(err)
	}
	blueprintData := strings.Split(string(content), "\n")
	blueprints := [][7]int{}
	for _, b := range blueprintData {
		if b == "" {
			continue
		}
		matches := re.FindAll([]byte(b), 7)
		if err != nil {
			panic(err)
		}

		b := [7]int{}
		for i, m := range matches {
			v, err := strconv.Atoi(string(m))
			if err != nil {
				panic(err)
			}
			b[i] = int(v)
		}
		blueprints = append(blueprints, b)
	}
	// max_geode := use_blueprint(blueprints[1])
	// fmt.Println("Finished with", 2, ":", max_geode)

	i := 0
	result := 1
	for i < 3 {
		max_geode := use_blueprint(blueprints[i])
		fmt.Println("Finished with", i, ":", max_geode)
		result = result * max_geode

		i += 1
	}
	fmt.Println("Solution part 2:", result)
}

// T 31
// 4320 too low  (24*18*10)
// 5184 too low (24*18*12)

// T 32
// 9338 too low (29*23*14)
